index_votes.py

```diff
@@ -0,0 +1,118 @@
+    with pg_conn.cursor() as cur:
+        for event in events_to_process(
+            cur,
+            "votes",
+        ):
+            (
+                type,
+                block_height,
+                tx_idx,
+                msg_idx,
+                _,
+                _,
+                chain_num,
+                timestamp,
+                tx_hash,
+            ) = event[0]
+            normalize = {}
+            normalize["type"] = type
+            normalize["block_height"] = block_height
+            normalize["tx_idx"] = tx_idx
+            normalize["msg_idx"] = msg_idx
+            normalize["chain_num"] = chain_num
+            normalize["timestamp"] = timestamp
+            normalize["tx_hash"] = tx_hash
+            for entry in event:
+                (_, _, _, _, key, value, _, _, _) = entry
+                value = value.strip('"')
+                normalize[key] = value
+            logger.debug(normalize)
+            votes = fetch_votes_by_proposal(
+                normalize["block_height"], normalize["proposal_id"]
+            )
+            logger.info(votes)
+            insert_text = textwrap.dedent(
+                """
+            INSERT INTO votes (
+                type,
+                block_height,
+                tx_idx,
+                msg_idx,
+                chain_num,
+                timestamp,
+                tx_hash,
+                proposal_id,
+                voter,
+                option,
+                metadata,
+                submit_time
+            ) VALUES (
+                %s,
+                %s,
+                %s,
+                %s,
+                %s,
+                %s,
+                %s,
+                %s,
+                %s,
+                %s,
+                %s,
+                %s
+            );"""
+            ).strip("\n")
+            with pg_conn.cursor() as _cur:
+                for vote in votes:
+                    try:
+                        row = (
+                            normalize["type"],
+                            normalize["block_height"],
+                            normalize["tx_idx"],
+                            normalize["msg_idx"],
+                            normalize["chain_num"],
+                            normalize["timestamp"],
+                            normalize["tx_hash"],
+                            normalize["proposal_id"],
+                            vote["voter"],
+                            vote["option"],
+                            vote["metadata"],
+                            vote["submit_time"],
+                        )
+                        _cur.execute(
+                            insert_text,
+                            row,
+                        )
+                        logger.debug(_cur.statusmessage)
+                        pg_conn.commit()
+                        logger.info("vote inserted..")
+                    except ForeignKeyViolation as exc:
```

Comment:
this exception block handles the scenario where a proposal has not yet been pruned, but a vote has been cast. once that proposal is pruned and the indexer has recorded it in the proposals table, then the next time the indexer checks for votes, it will pick these up and insert them into the database for the proposal.

in other words, until we've indexed the proposal, we will not index the votes. the proposal is not considered historical until it is pruned, and so it does seem to make sense. votes are only considered historical once the proposal becomes historic.

that said, we can change this behavior, however if we do, we lose the ability to traverse the relationship via graphql as mentioned above (since we would need to drop the foreign key).

Reply:
> So currently we are fetching all votes each time a single vote is submitted, looping through the votes each time we fetch all votes, and then rolling back each commit if the commit fails due to a foreign key violation (i.e. the proposal has not yet been indexed because we have a constraint in place that requires the proposal to exist).

this works as follows:
1. find all `EventVote`'s that we have not yet indexed.
2. loop through those votes
3. for each of those votes, attempt to insert and commit the vote into the database 
4. if we encounter a foreign key error, it means the corresponding proposal has not yet been indexed
5. we only rollback the transaction for a single vote if 4 is hit

So votes that have their corresponding proposals in the database will indeed be committed.

> Wouldn't this require an EventVote to be emitted after the proposal is indexed but a proposal is only indexed after the proposal is pruned and the voting period has ended? 

It works because the process will always attempt to index votes which have not been successfully indexed.
We will keep trying to index votes over and over again, until it succeeds.
Until the proposal is pruned, we might have a queue of votes waiting for indexing.
As soon as the proposal is pruned and exists in the indexer database, all of the votes for that proposal will be indexed in the next run of the index votes process.

Reply:
Are we attempting to iterate through all votes from the response even after the first vote was not stored due to the foreign key violation with the constraint? I tested with a single member and vote but haven't tested with multiple yet.

Reply:
This approach is a bit concerning looking at the database logs:
https://github.com/regen-network/indexer/actions/runs/6042565087/job/16397918758?pr=41#step:5:88

We are attempting to index votes every second until the proposal has been indexed...

Reply:
Also, is this process stuck on an event for all groups and all proposals? For example, I have a group proposal in one group with a 2 week voting period and a group proposal in another group with a 1 day voting period. If I submit a vote to the group proposal with a 2 week voting period, then a vote to the group with a 1 day voting period, will my vote not be indexed for the group with the 1 day voting period until the first proposal has been indexed?

Reply:
I can try to test this out and answer my own questions but let me know if you have any thoughts in the meantime.

Reply:
> This approach is a bit concerning looking at the database logs:
> https://github.com/regen-network/indexer/actions/runs/6042565087/job/16397918758?pr=41#step:5:88
> 
> We are attempting to index votes every second until the proposal has been indexed...

So we can adjust the timing...? I'd be concerned if we had a proven issue with cpu/memory in the database

Reply:
>Also, is this process stuck on an event for all groups and all proposals? For example, I have a group proposal in one group with a 2 week voting period and a group proposal in another group with a 1 day voting period. If I submit a vote to the group proposal with a 2 week voting period, then a vote to the group with a 1 day voting period, will my vote not be indexed for the group with the 1 day voting period until the 2 week proposal has been indexed?

No, votes are only blocked if their corresponding proposal has not been pruned.

Reply:
> Are we attempting to iterate through all votes from the response even after the first vote was not stored due to the foreign key violation with the constraint? I tested with a single member and vote but haven't tested with multiple yet. We should make sure we are exiting the loop as all other votes for the proposal would fail and rollback.

Makes sense, we can do that easily.

Reply:
> I tested with a single member and vote but haven't tested with multiple yet. 

I think I may have a bug in the way the indexer works for multiple voters.
I'll look into next week.
But basically I think I missed the fact that this API will return all the votes on the proposal including the voter for a particular `EventVote`:
/cosmos/group/v1/votes_by_proposal/{proposal_id}
So I think we'd end up with duplicates in the database with the current code.

I'll figure this out next week and do some more testing.
Thanks for mentioning multiple voters @ryanchristo .

Reply:
Yea, just posted this at the same time: https://github.com/regen-network/indexer/pull/38#discussion_r1312389707